### PR TITLE
Add manual section on method design and more advanced ambiguity resolution

### PR DESCRIPTION
**Pull request by timholy**
I think many people struggle with how to resolve ambiguities, and this really gets into how one should design Method hierarchies in Julia. For some time I've thought we could use a bit more documentation on this topic; here's an attempt to start that conversation.

There are many caveats to this:
- it's surely incomplete
- it ventures opinions that may or may not be defensible
- it may not be in the right place (it may make use of some concepts not yet introduced)
- it may not strike the right balance (it has a feel somewhere between "high level design patterns" and something more along the lines of the cookbook approach of "performance tips")

In other words, it's just about perfect for generating discussion on the topic :wink:.

### CODE REVIEW

**Comment by musm** [comment]
Thanks.

What's your opinion on including some of the following points or thoughts.
From steveng
`my understanding is that if you write a function f(T::Type), then T is just a value that is determined at runtime (the same version of f is compiled for all values of T), whereas if you write f{T}(::Type{T}) then T is part of the type signature of the function — hence it is known at compile time and a specialized version of f is compiled for every T.` Ref 17794

This statement only apply to types if my understanding is correct? In other words, there is no difference between the following two functions
`fun1{T<:Number}(x::T) = x`
`fun2(x::Number) = x`

File: doc/manual/methods.rst [hunk 1]
```diff
@@ -552,4 +558,164 @@ arguments::
...
+will generate ambiguities for anyone who defines a method
+::
+
+   -{T}(A::MyArrayType{T}, b::T)
+
+In order of preference, some strategies are:
+
+- avoid defining either of these methods: instead, rely on a generic method ``-{T}(A::AbstractArray{T}, b)`` and make sure this method is implemented with generic calls (like ``similar`` and ``-``) that do the right thing for each container type and element type *separately*.
+
+- if at least one of these must be defined, prioritize dispatch on the container
```
Comment (nalimilan) [inline doc/manual/methods.rst]: I don't understand what you mean by this point.

Comment (timholy) [inline doc/manual/methods.rst]: That is to say, where possible don't dispatch on `T`, but it's OK (often necessary) to dispatch on `MyArray`.

**Comment by timholy** [comment]
@musm, that's right. The limitation regarding types (`T::Type` vs `::Type{T}`) may be more about where Julia is now than anything truly fundamental; in other words, I can easily see it getting to the point where it doesn't matter for types, either.

File: doc/manual/methods.rst [hunk 2]
```diff
@@ -552,4 +558,164 @@ arguments::
...
+   f(x::Int, y) = 2
+
+by defining a method
+::
+
+   f(x::Int, y::Int) = 3
+
+This is often the right strategy; however, there are circumstances
+where following this advice blindly can be counterproductive. In
+particular, the more methods you have, the more possibilities there
```
Comment (mauro3) [inline doc/manual/methods.rst]: Just to be on the save side, I'd write "...,the more methods a generic function has,..."

File: doc/manual/methods.rst [hunk 3]
```diff
@@ -552,4 +558,164 @@ arguments::
...
+
+will generate ambiguities for anyone who defines a method
+::
+
+   -{T}(A::MyArrayType{T}, b::T)
+
+In order of preference, some strategies are:
+
+- avoid defining either of these methods: instead, rely on a generic method ``-{T}(A::AbstractArray{T}, b)`` and make sure this method is implemented with generic calls (like ``similar`` and ``-``) that do the right thing for each container type and element type *separately*.
+
```
Comment (mauro3) [inline doc/manual/methods.rst]: spurious new-line?

File: doc/manual/methods.rst (hunk 1 above)
Comment (mauro3) [inline doc/manual/methods.rst]: Note quite clear to me what you mean, maybe add example?

**Comment by mauro3** [comment]
Good addition, thanks for writing down your insights.

**Comment by Sacha0** [comment]
The material in this pull request is insightful. Any hope of reviving it? Thanks!

**Comment by timholy** [comment]
Comments address and rebased.

**Comment by tkelman** [comment]
the manual isn't in rst any more, bad rebase adding that file

**Comment by timholy** [comment]
Indeed. See #20702.

File: doc/src/manual/methods.md [hunk 4]
```diff
@@ -629,4 +632,209 @@ function emptyfunc
...
+
+The better design would be to define your call hierarchy like this:
+
+```julia
+struct NoPad end  # indicate that no padding is desired, or that it's already applied
+
+myfilter(A, kernel) = myfilter(A, kernel, Replicate())  # default boundary conditions
+
+function myfilter(A, kernel, ::Replicate)
+    Apadded = pad(A, kernel)
```
Comment (tkelman) [inline doc/src/manual/methods.md]: would it make more sense for this to be the same padding function `replicate_edges(A, size(kernel))` as above?
